src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -54,10 +67,199 @@ import qualified Data.ByteString.Short.Internal as SBS
 import           Data.Maybe (fromMaybe, isJust)
 import           Data.Monoid (Dual (..))
 import           Data.Primitive.ByteArray
+import           Data.STRef
 import           Data.Word
+import           Database.LSMTree.Internal.Run.BloomFilter (Bloom, MBloom)
+import qualified Database.LSMTree.Internal.Run.BloomFilter as Bloom
+import           Database.LSMTree.Internal.Run.Index.Compact (Chunk,
+                     CompactIndex, MCompactIndex)
+import qualified Database.LSMTree.Internal.Run.Index.Compact as Index
 import qualified Database.LSMTree.Internal.Serialise as S
 import           Foreign.Ptr
 
+{------------------------------------------------------------------------------
+  TODO: placeholders until #55 is merged
+-------------------------------------------------------------------------------}
+
+data Append =
+    -- | One or more keys are in this page, and their values fit within a single
+    -- page.
+    AppendSinglePage S.SerialisedKey S.SerialisedKey
+    -- | There is only one key in this page, and it's value does not fit within
+    -- a single page.
+  | AppendMultiPage S.SerialisedKey Word32 -- ^ Number of overflow pages
+
+-- | Append a new page entry to a mutable compact index.
+append :: Append -> MCompactIndex s -> ST s [Chunk]
+append = error "append: replace by Database.LSMTree.Internal.Run.Index.Compact.append"
+
+mkAppend :: PageAcc -> Append
+mkAppend p = case unStricterList $ pageKeys p of
+    []                     -> error "mkAppend: empty list"
+    [k] | numBytes <= 4096 -> AppendSinglePage k k
+        | otherwise        -> AppendMultiPage  k (fromIntegral $ numBytes `rem` 4096 - 1)
+    ks                     -> AppendSinglePage (last ks) (head ks)
+  where
+    numBytes = pageSizeNumBytes $ pageSize p
+
+{-------------------------------------------------------------------------------
+  Incremental, in-memory run construction
+-------------------------------------------------------------------------------}
+
+-- | A mutable structure that keeps track of incremental, in-memory run
+-- construction.
+--
+-- Use 'new' to start run construction, add new key\/operation pairs to the run
+-- by using 'add' and co, and complete run construction using 'finalise'.
+data MRun s = MRun {
+      mbloom         :: !(MBloom s RawKey)
+    , mindex         :: !(MCompactIndex s)
+      -- | Used to construct the full CompactIndex when run construction is
+      -- completed.
+    , indexChunksRef :: !(STRef s [[Index.Chunk]]) -- TODO: make stricter
+    , currentPageRef :: !(STRef s PageAcc)
+    }
+
+-- | @'new' npages@ starts an incremental run construction.
+--
+-- @npages@ should be an upper bound on the number of pages that will be yielded
```

Comment:
Should this really be an upper bound? It's only used for tuning filter and index, so I imagine giving an estimate that ends up being a little too low is better than one that's way too large?

Reply:
Ah, absolutely. That makes a lot more sense.